pkg/controllers/managementagent/monitoring/operatorHandler.go

```diff
@@ -40,32 +40,37 @@ func (h *operatorHandler) syncCluster(key string, obj *mgmtv3.Cluster) (runtime.
 		return obj, nil
 	}
 
-	var newCluster *mgmtv3.Cluster
-	var err error
+	cpy := obj.DeepCopy()
+	var err, updateErr error
 	//should deploy
 	if obj.Spec.EnableClusterAlerting || obj.Spec.EnableClusterMonitoring {
-		newObj, err := v32.ClusterConditionPrometheusOperatorDeployed.Do(obj, func() (runtime.Object, error) {
-			cpy := obj.DeepCopy()
+		newObj, err := v32.ClusterConditionPrometheusOperatorDeployed.Do(cpy, func() (runtime.Object, error) {
 			return cpy, deploySystemMonitor(cpy, h.app, h.catalogManager, h.clusterName)
 		})
 		if err != nil {
 			logrus.Warnf("deploy prometheus operator error, %v", err)
 		}
-		newCluster = newObj.(*mgmtv3.Cluster)
+		cpy = newObj.(*mgmtv3.Cluster)
 	} else { // should withdraw
-		newCluster = obj.DeepCopy()
-		if err = withdrawSystemMonitor(newCluster, h.app); err != nil {
+		if err = withdrawSystemMonitor(cpy, h.app); err != nil {
 			logrus.Warnf("withdraw prometheus operator error, %v", err)
 		}
 	}
 
-	if newCluster != nil && !reflect.DeepEqual(newCluster, obj) {
-		if newCluster, err = h.clusters.Update(newCluster); err != nil {
-			return nil, err
-		}
-		return newCluster, nil
+	if !reflect.DeepEqual(cpy, obj) {
+		cpy, updateErr = h.clusters.Update(cpy)
+		return cpy, nil
 	}
-	return obj, nil
+
+	if err != nil {
+		return nil, err
+	}
+
+	if updateErr != nil {
+		return nil, updateErr
+	}
+
+	return cpy, nil
```

Comment:
Instead of doing the error checking at the end, we should check and return the error when the error is created (e.g. replace the `logrus.Warnf(args)` with `return nil, fmt.Errorf(args)`).

The framework for controllers should automatically take care of requeuing the cluster and triggering the handlers (including this one) when an error is returned.

Reply:
I think the old code doesn't return err directly because it needs to update the error reason and status to the cluster object, so I check error after update cluster.

Reply:
The code you linked seems to be using `_, updateErr = cd.clusters.Update(cluster)` since it wants to return the error created by `Update` at the end. However, here it seems like the `updateErr` is being entirely discarded either way (since you call `return cpy, nil` right after the `updateErr` ever gets a non-nil value).